### DISCUSSION TITLE
Unfulfillable offer pattern

### DISCUSSION THREAD

**Discussion by carlos-kryha** [Agoric/agoric-sdk#7274] [Q&A] [answered]
Having experienced the convenience of Zoe and Offer-safety for smart contract development, I find myself attempting to use it whenever possible. But one particularly common use-case requires using offers in a way that (at least to me) feels "hacky". I want to start a discussion around the pattern and tackle the following questions:

1. Is it advisable to use offers in this way? (as described in Current Workaround)
2. Are there any alternatives that achieve the same functionality?
3. Is this functionality something that future versions of Zoe offer a custom solution for?

# Use-case
Allow users to prove ownership of ERTP assets to a smart contract while being certain the contract won't take the assets from the user

# Current limitations
From a user's pov, escrowing assets on Zoe using an Offer implies the intention of giving them away, but in this case the user wants only to show them. One way to go about it would be to match the `want` and `give` properties to indicate that you're willing to give your assets but want them back, but Zoe does not allow assets of the same brand to be both in the `give` and `want` of the same offer

# Current workaround
Instead of requesting the same assets back in the offer's `want`, if you simply make up a fictional amount and include it in your `want`,  you can be certain the offer will never fulfill and by exiting the offer you can get your assets back

# Security Considerations 
If the offer were to be set up to disallow users from exiting, the assets could be locked up in Zoe

# Thoughts
Although the workaround works well per my testing, it feels like I'm using offer safety in a way that it wasn't meant to be used, and adds overhead when explaining why you would form an offer in such a way. It would be nice if on top of an offer's `give` and `want`, there was a third option `show` which escrowed the assets on Zoe and allowed the contract to inspect them without the possibility to reallocate.

I'd love to hear your thoughts on this

**Comment by Chris-Hibbert** [comment, marked as answer]
We have explicitly talked about this possibility ( `{ want: 3 unobtanium }` ) and it seems perfectly reasonable. `show` is a plausible alternative, but one we hadn't contemplated before.

In order for this to be helpful, a contract would have to be written to expect that it should assay the want, but not expect to reallocate it. Contracts set up this way might experiment with different ways of expressing the unmatchable `want`. They might have their own issuer that allows amounts but never mints anything, or they might just ignore the `want` and leave it up to the client to create a relevant mint.

We experimented with similar techniques (assaying goods) in order to support voting schemes based on ownership of shares. For voting, we also wanted to ensure that only one vote could be cast per share, but there are probably cases where repeated use of the same asset don't matter as much.

**Comment by dtribble** [comment]
I've advocated the current pattern in a few cases, but another limitation is that you cannot have a thing that you show, while then allowing a reallocate among give/want. The `show` case would provide some general support for "credentials".

However an important limitation of `show` would be that you cannot have more than one offer in flight with the same creds.  That is a problem e.g., for the voting that @Chris-Hibbert mentioned.  So instead, you could ask the precious asset for an expendable attestation. For voting, that would be "get a newly minted 'vote in election X' which is consumable".  Then that token would be verified and consumed in the offer. Could you do something similar?

Two notes wrt the security consideration: 

* the submitter of the offer gets to decide whether they use any exit criteria other than `onDemand`, so they cannot get tricked into locking up their asset. That of course means the pattern shouldn't be use in contract that require lock-ups in offers.
* it is not ideal that a mistake (forgetting to add the unobtanium want) could put the user's precious asset at risk
* for using the pattern in the meantime, it may be useful to have a canonical brand with a burnt mint so that it's clear that it's not possible to mint the unobtanium, and that's clear in the brand specification of the offer.

**Reply by carlos-kryha** [reply]
"Credentials" would be a good way to describe what we're after here. Our use-case does not require using a single credential in multiple active offers (feel free to check out the comment below for more details on the use-case), but I'm curious, the pattern you're suggesting would be very similar to the continuing invitation pattern, right?

I think having a cannonical brand for this would be great, could make it a bit more declarative and easier to use.

**Comment by carlos-kryha** [comment]
Great! Glad to hear this would be an acceptable use of offer safety. The voting scenario is also interesting, maybe I can explain our use-case for some context. We would like to use nfts to grant access to private content, here's a draft of the intended flow:
1. Content owner sends a set of nfts to the contract, along with a reference to the content they (the nfts) are meant to unlock
2. Contract registers the nfts as access control for the content and returns them to the content owner
3. Content owner distrubutes the nfts to users
4. Users _show_ the nfts to the contract in return for temporary access to the content they are tied to

There are two places where this pattern would be useful in this scenario:
1. Registering assets for access control (steps 1, 2): since there is no reallocation here, we could simply use a regular contract call rather than an invitation, but then we would have no way of knowing if the caller owns the nft or simply has access to its content
2. Consuming content as the owner of an nft that has been configured as access control (step 3): here a user need only prove they own the nft in order to get [temporary] access to the content

For this use-case we want to use nfts as you would a physical key, meaning a single key can only be used in one place at once, yet there is no limit to the amount of times it can be used. Should the content owner want multiple users to access the content simultaniously, they would need to register multiple keys (nfts)
